bluetooth.go

```diff
@@ -0,0 +1,367 @@
+	if err != nil {
+		return err
+	}
+	err = d.bt_init_buffers()
+	if err != nil {
+		return err
+	}
+	err = d.bt_wait_awake()
+	if err != nil {
+		return err
+	}
+	err = d.bt_set_host_ready()
+	if err != nil {
+		return err
+	}
+	d.bt_toggle_intr()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Device) bt_upload_firmware(firmware string) error {
+	d.trace("bt_init:start")
+	versionlength := firmware[0]
+	if versionlength > 8 {
+		return errors.New("invalid version length")
+	}
+	var version uint32
+	for i := 0; i < int(versionlength); i++ {
+		version |= uint32(firmware[i]) << i
+	}
+	// Skip version + 1 extra byte as per cybt_shared_bus_driver.c
+	firmware = firmware[versionlength+2:]
+	// buffers
+	rawbuffer := u32AsU8(d._sendIoctlBuf[:])
+	alignedDataBuffer := rawbuffer[:256]
+	btfwCB := firmware
+	hfd := hexFileData{
+		addrmode: whd.BTFW_ADDR_MODE_EXTENDED,
+	}
+	var auxbuf [4]byte
+	for {
+		numFwBytes := bt_read_firmware_patch_line(btfwCB, &hfd)
+		if numFwBytes == 0 {
+			break
+		}
+		fwBytes := hfd.ds[:numFwBytes]
+		dstStartAddr := hfd.dstAddr + whd.CYW_BT_BASE_ADDRESS
+		var alignedDataBufferIdx uint32
+		if !isaligned(dstStartAddr, 4) {
+			// Pad with bytes already in memory.
+			numPadBytes := dstStartAddr % 4
+			paddedDstStartAddr := aligndown(dstStartAddr, 4)
+			memoryValue, _ := d.bp_read32(paddedDstStartAddr)
+
+			_busOrder.PutUint32(auxbuf[:], memoryValue)
+			for i := 0; i < int(numPadBytes); i++ {
+				alignedDataBuffer[alignedDataBufferIdx] = auxbuf[i]
+				alignedDataBufferIdx++
+			}
+
+			dstStartAddr = paddedDstStartAddr
+		}
+		// Copy firmware bytes after padding bytes.
+		alignedDataBufferIdx += uint32(copy(alignedDataBuffer[alignedDataBufferIdx:], fwBytes))
+
+		dstEndAddr := dstStartAddr + alignedDataBufferIdx
+		if !isaligned(dstEndAddr, 4) {
+			offset := dstEndAddr % 4
+			numPadBytesEnd := 4 - offset
+			paddedDstEndAddr := aligndown(dstEndAddr, 4)
+			memoryValue, _ := d.bp_read32(paddedDstEndAddr)
+			_busOrder.PutUint32(auxbuf[:], memoryValue)
+			for i := offset; i < 4; i++ {
+				alignedDataBuffer[alignedDataBufferIdx] = auxbuf[i]
+				alignedDataBufferIdx++
+			}
+			dstEndAddr += numPadBytesEnd
+		}
+		bufferToWrite := alignedDataBuffer[0:alignedDataBufferIdx]
+		if dstStartAddr%4 != 0 || dstEndAddr%4 != 0 || alignedDataBufferIdx%4 != 0 {
+			return errors.New("unaligned BT firmware bug")
+		}
+		const chunksize = 64 // Is writing in 64 byte chunks needed?
+		numChunks := align(alignedDataBufferIdx, 64)
+		for i := uint32(0); i < numChunks; i += chunksize {
+			offset := i * chunksize
+			end := (i + 1) * chunksize
+			if end > alignedDataBufferIdx {
+				end = alignedDataBufferIdx
+			}
+			chunk := bufferToWrite[offset:end]
+			err := d.bp_write(dstStartAddr+offset, chunk)
+			if err != nil {
+				return err
+			}
+			time.Sleep(time.Millisecond) // TODO: is this sleep needed?
+		}
+	}
+	return nil
+}
+
+func (d *Device) hci_set_read_handler(fn func(b []byte) error) {
+
```

Comment:
Is this supposed to be called from `RecvHCIHandle()` or was it left in by accident?

Reply:
Ah, that was left in by accident. Thanks for catching it!